docs/AsyncWrap/README.md

```diff
@@ -1,15 +1,202 @@
+of tracing the latter is usually what is meant.
 
-An older-ish intro to AsyncWrap is available in Trevor Norris's blog post
-[AsyncWrap Tutorial - An Introduction](http://blog.trevnorris.com/2015/02/asyncwrap-tutorial-introduction.html).
+_The reasoning for the current naming confusion, is that the API part implements
+the hooks through the AsyncWrap class. But there is nothing inherently correct
+in doing that, for example if v8 provided those facilities the AsyncWrap class
+would not need be involved in the AsyncWrap API._
 
-There was also some discussion in [issue #21](https://github.com/nodejs/tracing-wg/issues/21#issuecomment-142727693).
+For the remaining description the API part is what is meant by AsyncWrap.
+
+## Handle objects
+
+AsyncWrap emits events (hooks) that informs the consumer about the life of all handle
+objects in node. Thus in order to understand AsyncWrap one must first understand
+handle objects.
+
+The API that node exposes is mostly defined in JavaScript files. However
+ECMAScript does not define any API for creating a TCP socket, reading a file
+etc.. That logic is implemented in C++ using libuv and the v8 API. The JavaScript in
+nodecore interacts with this C++ layer using the handle objects.
+
+For example to create and connect a TCP socket:
+
+```javascript
+const TCP = process.binding('tcp_wrap').TCP;
+const TCPConnectWrap = process.binding('tcp_wrap').TCPConnectWrap;
+
+const req = new TCPConnectWrap();
+req.oncomplete = oncomplete;
+req.address = address;
+req.port = port;
+
+const socket = new TCP();
+socket.onread = onread;
+socket.connect(req, address, port);
+```
+
+This example uses two handle objects. The first one (`TCPConnectWrap`) is for
+connecting the socket, the second one (`TCP`) is for maintaining the connection.
+
+`TCPConnectWrap` gets its information by setting properties on the handle
+object, like `address` and `port`. Those properties are read by the C++ layer,
+but can also be inspected from the AsyncWrap hooks. When the handle is create
+using `new TCPConnectWrap()` the `init` hook is called.
+
+A `oncomplete` property is also set, this is the callback for when the connection is made or failed. Just before calling `oncomplete` the `before` hook
+is called, just after the `after` hook is called.
+
+The `TCP` handle works exactly the same way, except that the information
+is passed as arguments to a method `.connect` and the `onread` function
+is called multiply times, thus it behaves like an event. This also means that
+the `before` and `after` hooks are called multiply times.
+
+Thus one should expect the hooks be called in the following order:
+
+```javascript
+init // TCPConnectWrap
+init // TCP
+=== tick ===
+before // TCPConnectWrap
+after // TCPConnectWrap
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+before // TCP
+after // TCP
+=== tick ===
+...
+```
+
+_tick_ indicates there is at least one tick between the text above and the text below.
```

Comment:
The word "tick" is unfortunately ambiguous. Do you mean in the sense of `nextTick()` or as in a tick of the event loop?

One thing that isn't clear from this is which operation generated which object. For example. `net.createServer()` will create a `TCP`, and `net.connect()` will create the `TCPConnectWrap`, but another `TCP` is created that is the actual connection to the server. This will have it's own init callback called.

Part of this behavior is pending on https://github.com/nodejs/node/pull/3216, which operates more like I had intended it to from the beginning. Log from the callbacks of that test to see what I mean.